plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/MachineProviderImplTest.java

```diff
@@ -697,6 +697,56 @@ public void shouldBindCommonAndDevVolumesToContainerOnDevInstanceCreationFromRec
         assertEquals(actualBinds.length, expectedVolumes.size());
         assertEquals(new HashSet<>(asList(actualBinds)), new HashSet<>(expectedVolumes));
     }
+    
+    @Test
+    public void shouldBindVolumesFromToContainerOnDevInstanceCreationFromRecipe() throws Exception {
+    	
+    	List<String> cheServerVolumes = new ArrayList<>(asList("SomeVolumeName:/data/lib"));
+    	String expectedVolumesFrom = "che-server";
+    	
+    	provider = new MachineProviderBuilder().setVolumesFrom(expectedVolumesFrom)
+    									       .build();
+    	
+    	final boolean isDev = true;
+    	
+    	CheServiceImpl service = createService();
+    	service.setContainerName(expectedVolumesFrom);
+    	service.setVolumes(cheServerVolumes);
+    	createInstanceFromRecipe(isDev, service);
+    	
+        ArgumentCaptor<CreateContainerParams> argumentCaptor = ArgumentCaptor.forClass(CreateContainerParams.class);
+        verify(dockerConnector).createContainer(argumentCaptor.capture());
+        verify(dockerConnector).startContainer(any(StartContainerParams.class));
+        
+        final String[] actualVolumesFrom = argumentCaptor.getValue().getContainerConfig().getHostConfig().getVolumesFrom();
+        assertTrue(Arrays.asList(actualVolumesFrom).contains(expectedVolumesFrom));
+    }
+    
+    @Test
+    public void shouldBindVolumesFromToContainerOnDevInstanceCreationFromRecipeWhenServiceHasVolumesFromSet() throws Exception {
+    	
+    	List<String> cheServerVolumes = new ArrayList<>(asList("SomeVolumeName:/data/lib"));
+    	List<String> cheServerVolumesFrom = new ArrayList<>(asList("SomeContainerName"));
+    	String expectedVolumesFrom = "che-server";
+    	
+    	provider = new MachineProviderBuilder().setVolumesFrom(expectedVolumesFrom)
+    									       .build();
+    	
+    	final boolean isDev = true;
+    	
+    	CheServiceImpl service = createService();
+    	service.setContainerName(expectedVolumesFrom);
+    	service.setVolumes(cheServerVolumes);
+    	service.setVolumesFrom(cheServerVolumesFrom);
+    	createInstanceFromRecipe(isDev, service);
+    	
+        ArgumentCaptor<CreateContainerParams> argumentCaptor = ArgumentCaptor.forClass(CreateContainerParams.class);
+        verify(dockerConnector).createContainer(argumentCaptor.capture());
+        verify(dockerConnector).startContainer(any(StartContainerParams.class));
+        
+        final String[] actualVolumesFrom = argumentCaptor.getValue().getContainerConfig().getHostConfig().getVolumesFrom();
+        assertTrue(Arrays.asList(actualVolumesFrom).contains(expectedVolumesFrom));
```

Comment:
Hi @garagatyi , perhaps my understanding is off so please correct me, but doesn't **CheServiceImpl** represent the Che service, while the **MachineProviderBuilder** provides the machine (i.e. workspace) instances? Testing for an assertion that **.getHostConfig().getVolumesFrom()** returns a length of 2 causes the test to fail, so it is not the case that a **volumesFrom** is being added twice.

You asked me earlier if I had any suggestions for making the tests easier to understand - perhaps some kind of documentation that explained line by line what was happening in the test would be helpful.  As it stands, there seems to be a lot going on that I'm a bit unclear about, which means all I can do is make assumptions about what is happening based on the method and variable names.  I think some docs would help to reduce this confusion.